Make RosTeamWorkspace independent of installed path

mamueluth opened the pull request:
RTW should now work regardless of where it was cloned.
**Important:** The .ros_team_ws_rc has changed, DEFAULT_ROS_DISTRO="rolling" DEFAULT_ROS_VERSION=2 are not needed anymore, an alternative path location system was introduced

- How ros distro/version is selected was changed
- Add alternative path for ros location (ALTERNATIVE_ROS_ROLLING_LOCATION)
- Add setup script
- Some minor improvements in user input handling

Please make a qick check if working correctly!

destogl commented on scripts/_RosTeamWs_Defines.bash:
```diff
@@ -237,69 +279,111 @@ function print_and_exit {
+
+  if [ ! -d "/opt/ros/$ros_distro" ]; then
+    local upper_case=$(echo $ros_distro | tr '[:lower:]' '[:upper:]')
+    local alternative_ros_location=ALTERNATIVE_ROS_${upper_case}_LOCATION
```
Why do we need an alternative ROS location? We are actually interested in alternative RosTeamWorkspace locations. Can you give me a short example when this functionality is needed?

mamueluth replied:
If someone has installed ros at a different location than `/opt/ros/` (e.g. source install) we check if the requested ros distro is installed there. Otherwise, we cannot create a new package for example since this depends on `ros2 pkg create`.

mamueluth replied:
If make rtw independent from ros in the next step it could be possible that we don't need this anymore.

destogl replied:
OK, we can leave it, this is good!

destogl commented on scripts/create-new-package.bash:
```diff
@@ -22,20 +22,36 @@ echo -e "${TERMINAL_COLOR_USER_NOTICE}Your path is `pwd`. Is this your workspace
+do
+  case "$META" in
+        standard)
+            META="1"
```
can we keep "0", "1", "2"? It somehow more natural that "0" is default

destogl replied:
This should be actually default, so we only get to check "1" and "2" here.

mamueluth replied:
going to look if the default value in select can be changed

mamueluth replied:
Should the other select also be changed?

mamueluth replied:
since 1 is the default for select i don't change this.

destogl commented on scripts/os_configure/setup_auto_sourcing.bash:
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+#
+# Copyright 2021 Denis Stogl (Stogl Robotics Consulting)
```
```suggestion
# Copyright 2022 Manuel Muth (Stogl Robotics Consulting)
```

mamueluth replied:
Done

destogl commented on scripts/os_configure/setup_auto_sourcing.bash:
```diff
@@ -0,0 +1,68 @@
+sed -i "s|source <PATH TO ros_team_workspace>/setup.bash|source $FRAMEWORK_BASE_PATH/setup.bash|g" "$rtw_file_location"
+
+echo ""
+user_decision "Do you want to add add colored output to your shell? This will show the current git branche, ros workspace and time in terminal. If you are unsure you can test this later by uncommenting the \"export PS1=\" in the .ros_team_ws_rc file or rerun the setup-auto-sourcing command."
```
```suggestion
user_decision "Do you want to add add colored output to your shell? This will show the current git branch, ros workspace and time in terminal. If you are unsure you can test this later by uncommenting the \"export PS1=\" in the .ros_team_ws_rc file or rerun the setup-auto-sourcing command."
```

mamueluth replied:
Done

destogl commented on scripts/setup-repository.bash:
```diff
@@ -26,8 +26,8 @@ read -p "If so press <ENTER> otherwise <CTRL>+C and start the script again from
-source $script_own_dir/_RosTeamWs_Defines.bash
-check_ros_distro ${ROS_DISTRO}
+source $script_own_dir/../setup.bash
+check_ros_distr ${ROS_DISTRO}
```
```suggestion
check_and_set_ros_distro_and_version ${ROS_DISTRO}
```

mamueluth replied:
Done

destogl commented on scripts/os_configure/setup_auto_sourcing.bash:
```diff
@@ -0,0 +1,68 @@
+    export_ps1_line_number=$(grep -n "#.*export PS1=" .ros_team_ws_rc | grep -Eo '^[^:]+')
+    # uncomment
+    if ! [ -z "$export_ps1_line_number" ]; then
+        sed -i "${export_ps1_line_number},${export_ps1_line_number}s|#||g" "$rtw_file_location"
```
Should we move this `PS1` code into some external file of RosTeamWorkspace? So that we make .ros_team_ws_rc file shorter? What do you think?

mamueluth replied:
Done. Moved color definitions to `scripts/configuration.` and setup script to `scripts/.`

destogl reviewed (commented):
You really did a great work! Just a few minor comments

destogl commented on docs/tutorials/setting_up_rtw.rst:
````diff
@@ -16,32 +18,42 @@ Source the ``setup.bash``` in the top folder of RosTeamWorkspace:
-Execute the following to configure the RosTeamWorkspace permanently.
 
-**ATTENTION**: The following will work only if RosTeamWorkspace is checked out in ``/opt/RosTeamWS/ros_ws_rolling/src/`` folder! We are working on fixing this. The progress can se followed in `this issue <https://github.com/StoglRobotics/ros_team_workspace/issues/51>`_.
+This is going to configure the RosTeamWorkspace permanently. If you want to revert those changes or prefer to do them by yourself simply follow the next few steps:
````
```suggestion
This is going to configure the RosTeamWorkspace permanently. If you want to revert those changes or prefer to do them by yourself simply follow the next few steps.
```

destogl commented on scripts/setup_auto_sourcing.bash:
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
```
the script was not working when I already have had .ros_team_ws file in my workspace. why not back it up in `.ros_team_ws.bkp` file and create a new one. And also notify a user about it.

destogl replied:
Can you also please use colors when notifying user? Depending on what you want, confirmation of show information then we should use different colors. They are already defined in `_RosTeamWs_Defines` file.

mamueluth replied:
Done

mamueluth replied:
Did also fix the second (not here but in nextcloud mntioned) error with this [commit](https://github.com/StoglRobotics/ros_team_workspace/pull/55/commits/1baa8d1797150ebf416769bf7af42f176a5414bf)

destogl commented:
When I don't setup the rtw permanently and start executing `setup-ros-workspace`, I get following error messages:

```
cp: cannot stat '/home/deniss/.ros_team_ws_rc': No such file or directory
sed: can't read /home/deniss/.ros_team_ws_rc: No such file or directory
sed: can't read /home/deniss/.ros_team_ws_rc: No such file or directory
```

Should we first check if this files exist?

mamueluth commented:
- [ ]  Issue Nr. 1

- [ ] Issue Nr. 2
- Added conditional assignment of variables

- [ ] Issue Nr. 3
- [ ] Issue Nr. 4
- fixed it (used rtw_file instead of rtw_file_location, this was a mistake)
- [ ] Issue Nr. 5

destogl commented:
This works fine now!

destogl commented on docs/tutorials/setting_up_rtw.rst:
````diff
@@ -16,32 +18,42 @@ Source the ``setup.bash``` in the top folder of RosTeamWorkspace:
-- or :ref:`other use-cases <uc-index>`.
+- :ref:`setup your first workspace <uc-setup-workspace>`,
+- :ref:`quick-start <tutorial-quick-start>`,
+- or check out one of our:ref:`other use-cases <uc-index>`.
````
```suggestion
- or check out one of the :ref:`use-cases <uc-index>`.
```

destogl commented on scripts/configuration/terminal_coloring.bash:
```diff
@@ -0,0 +1,90 @@
+if [ -n "$SSH_CLIENT" ]; then text="-ssh-session"
+fi
+
+function parse_vc_branch_and_add_brackets {
```
```suggestion
function parse_git_branch_and_add_brackets {
```

destogl commented on scripts/configuration/terminal_coloring.bash:
```diff
@@ -0,0 +1,90 @@
+  fi
+}
+
+export PS1="\[${TERMINAL_COLOR_LIGHT_GRAY}\]"'[\t]\['"\[${TERMINAL_COLOR_LIGHT_GREEN}\]"'\u\['"\[${TERMINAL_COLOR_LIGHT_GRAY}\]"'@\['"\[${TERMINAL_COLOR_BROWN}\]"'\h\['"\[${TERMINAL_COLOR_YELLOW}\]"'${text}\['"\[${TERMINAL_COLOR_LIGHT_GRAY}\]"':'"\["'$(set_ros_workspace_color)'"\]"'$(parse_ros_workspace)\['"\[${TERMINAL_COLOR_GREEN}\]"'$(parse_vc_branch_and_add_brackets)>\['"\[${TERMINAL_COLOR_LI ...
```
```suggestion
export PS1="\[${TERMINAL_COLOR_LIGHT_GRAY}\]"'[\t]\['"\[${TERMINAL_COLOR_LIGHT_GREEN}\]"'\u\['"\[${TERMINAL_COLOR_LIGHT_GRAY}\]"'@\['"\[${TERMINAL_COLOR_BROWN}\]"'\h\['"\[${TERMINAL_COLOR_YELLOW}\]"'${text}\['"\[${TERMINAL_COLOR_LIGHT_GRAY}\]"':'"\["'$(set_ros_workspace_color)'"\]"'$(parse_ros_workspace)\['"\[${TERMINAL_COLOR_GREEN}\]"'$(parse_git_branch_and_add_brackets)>\['"\[${TERMINAL_COLOR_LIGHT_PURPLE}\]"'\W\['"\[${TERMINAL_COLOR_LIGHT_PURPLE}\]"'$\['"\[${TERMINAL_COLOR_NC}\]"'\[\e[m\] '
```

destogl reviewed (approved):
Thanks!
